string color
bool is_upper
int16[] range
